### PR TITLE
code loading fixes for cases where wrong version of package is loaded 

### LINKED ISSUE #45977: Wrong version of package can get up getting loaded if environment is not fully instantiated
Code loading contains two steps.:

- Identify the UUID of the package getting loaded. For example, if `Parsers` is loaded into Main, we check that there is a `[deps]` entry with `Parsers` in the current project and we find the UUID from there.
- Find the location of the package. This goes through all the environments and looks for a manifest that contains the UUID from above and computes the path.

The issue occurs when (for some reason) the source files of the package are not present for the package in the active environment. Code loading will then keep looking into manifests lower into the stacked env, and if it finds one, it breaks reproducibility. Here is an example:

```
using Pkg
using TOML
getversion(pkgpath) = TOML.parsefile(abspath(pkgpath, "..", "..", "Project.toml"))["version"]

# Add Parsers at version 1.0 in the global env
Pkg.add(name = "Parsers", version = "1.0")
@show getversion(Base.find_package("Parsers"))

# Add Parsers at version 2.0 in a temp env
Pkg.activate(; temp=true)
Pkg.add(name = "Parsers", version = "2.0")
@show getversion(Base.find_package("Parsers"))

# Now, remove the Parsers version 2.0 from the file system, this is equivalent to not having a fully instantiated manifest
# which `Pkg.status` shows:
rm(abspath(Base.find_package("Parsers"), "..", ".."); recursive=true)
Pkg.status()

# At this point, `find_package´ finds the Parsers.jl in the global environment which is at the wrong version
# Doing a `using Parsers` would load an inconsistent version with the current manifest.
@show getversion(Base.find_package("Parsers"))
using Parsers # now loads the wrong version
```

Note that this is not the same as when we load something like a dev-package from the shared env where it ends up loading dependencies at versions of the currently active environment. It is more the opposite.

So the bug here is that the package that ends up getting loaded depends on what happens to be installed in the depot.

### PR DESCRIPTION
**Pull request by KristofferC**
This fixes the two cases given in https://github.com/JuliaLang/julia/issues/45977 where the wrong package version gets loaded.

The first case is where the package version is not found in the depot. In that case, we now return a `missing` to signal not to look any further through the env stack.

The second case is where a manifest entry is not present for a package identified in the current env. This is fixed by returning the env a package got identified with and then stopping searching through envs in `locate_package` when that env is hit.

Fixes #45977

### CODE REVIEW

File: test/loading.jl [hunk 1]
```diff
@@ -920,3 +920,58 @@ end
...
+            uuid = "6801f525-dc68-44e8-a4e8-cabd286279e7"
+            git-tree-sha1 = "efc7e24c53d6a328011975294a2c75fed2f9800a"
+            """)
+
+        # This SHA does not exist in the depot.
+        write(joinpath(tmp, "Env1", "Manifest.toml"), """
+        [[Baz]]
+        uuid = "6801f525-dc68-44e8-a4e8-cabd286279e7"
+        git-tree-sha1 = "5f2f6e72d001b014b48b26ec462f3714c342e167"
+        """)
```
Comment (IanButterworth) [inline test/loading.jl:951]: Minor indenting 

```suggestion
            [[Baz]]
            uuid = "6801f525-dc68-44e8-a4e8-cabd286279e7"
            git-tree-sha1 = "5f2f6e72d001b014b48b26ec462f3714c342e167"
            """)
```

**Review by IanButterworth — APPROVED** [review]
The logic here seems good to me.

**Comment by KristofferC** [comment]
>  This is fixed by returning the env a package got identified with and then stopping searching through envs in locate_package when that env is hit.

Standard libraries are special snowflakes so I think they have to be excluded from this.

**Comment by DilumAluthge** [comment]
@KristofferC Can this be backported to Julia 1.8.1

**Comment by KristofferC** [comment]
Maybe 1.8.2, I want this to be on master for a while before going into an actual release.

**Comment by mortenpi** [comment]
@KristofferC The nightly tests for DataSets.jl started failing at one point, and I bisected it to this commit. Now, I am not sure if it indicates a bug here or not -- [the DataSets tests try to add a dummy package with `LOAD_PATH`](https://github.com/JuliaComputing/DataSets.jl/blob/d35f11e2c669a50744b52289269bfc21ea13753d/test/driver_autoload.jl#L3) and, since this commit, [the dummy package is unable to load the package that is in `Main`](https://github.com/JuliaComputing/DataSets.jl/actions/runs/3358926986/jobs/5566435933#step:6:170). It'd kinda weird, but I don't see why it should fail.

It should be easy to repro by taking the current latest commit https://github.com/JuliaComputing/DataSets.jl/commit/d35f11e2c669a50744b52289269bfc21ea13753d and just run the tests with Julia#master. You may need a clean depot too -- at some point when I was playing around with it, it looked like that might have made a difference.

**Comment by KristofferC** [comment]
Changing the `pushfirst!` to `push!` on the `LOAD_PATH` should probaly fix it. Alternatively, check in a manifest in the https://github.com/JuliaComputing/DataSets.jl/tree/master/test/drivers/DummyStorageBackends package. What this PR fixed is that a package cannot load a dependency that is below itself in the load path since that means that the top environment is no longer reproducible. When the DummyStorageBackends is in the top of the environment it needs to find its dependencies in an environment that is at the same level or above.

I must say that this whole strategy of ransacking the load path to look for packages to "automagically" load seems very brittle to me.

**Pull request merged**